File: android/src/main/java/com/sensitiveinfo/internal/metadata/MetadataManager.kt
package com.sensitiveinfo.internal.metadata

import com.margelo.nitro.sensitiveinfo.StorageMetadata
import com.sensitiveinfo.internal.storage.PersistedMetadata

/**
 * Interface for managing metadata operations on Android.
 *
 * Encapsulates all metadata encoding, decoding, and manipulation logic.
 * Follows Single Responsibility Principle by focusing only on metadata.
 *
 * @since 6.0.0
 */
interface MetadataManager {
  /**
   * Decode metadata from persisted format.
   *
   * @param metadata The StorageMetadata from storage
   * @return Decoded StorageMetadata or null if invalid
   * @throws Exception if decoding fails
   */
  suspend fun decodeMetadata(metadata: StorageMetadata?): StorageMetadata?

  /**
   * Encode metadata to persisted format.
   *
   * @param metadata The StorageMetadata to encode
   * @return Encoded StorageMetadata for storage
   * @throws Exception if encoding fails
   */
  suspend fun encodeMetadata(metadata: StorageMetadata): StorageMetadata

  /**
   * Create default metadata with common values.
   *
   * @param securityLevel Security level for this item
   * @param accessControl Access control policy
   * @param alias Encryption key alias
   * @return StorageMetadata with timestamp set
   */
  fun createMetadata(
    securityLevel: String,
    accessControl: String,
    alias: String
  ): StorageMetadata
}

File: ios/Internal/MetadataManager.swift
import Foundation

/// Protocol for managing metadata operations.
///
/// Encapsulates all metadata encoding, decoding, and manipulation logic.
/// Follows Single Responsibility Principle by focusing only on metadata.
///
/// @since 6.0.0
protocol MetadataManager {
  /// Decode metadata from raw bytes.
  ///
  /// - Parameters:
  ///   - data: Raw metadata bytes from Keychain
  /// - Returns: Decoded StorageMetadata or nil if invalid
  /// - Throws: RuntimeError if decoding fails
  func decodeMetadata(from data: Data?) throws -> StorageMetadata?

  /// Encode metadata to raw bytes.
  ///
  /// - Parameters:
  ///   - metadata: StorageMetadata to encode
  /// - Returns: Encoded metadata bytes for Keychain storage
  /// - Throws: RuntimeError if encoding fails
  func encodeMetadata(_ metadata: StorageMetadata) throws -> Data

  /// Create default metadata with common values.
  ///
  /// - Parameters:
  ///   - securityLevel: Security level for this item
  ///   - accessControl: Access control policy
  ///   - alias: Encryption key alias
  /// - Returns: StorageMetadata with timestamp set
  func createMetadata(
    securityLevel: SecurityLevel,
    accessControl: AccessControl,
    alias: String
  ) -> StorageMetadata
}
